### ISSUE TITLE
Twitch API v5 shutdown

### ISSUE THREAD

**Issue by apo86** [MrSprigster/Twitch.tv-on-XBMC#615]
So, I don't know if anyone has this on their radar, because I certainly didn't, but it looks like the addon is not going to work after end of the month unless some (many? most?) API calls are changed.

In fact it has already started with some scheduled downtimes that one or the other might have noticed Monday and today. As per Twitch's timeline those downtimes will increase until the final shutdown on the 28th:
https://blog.twitch.tv/en/2021/07/15/legacy-twitch-api-v5-shutdown-details-and-timeline/

I can try to begin figuring out what needs to be changed, but honestly I don't know if I'm up to the task. And I also don't know if this even is a "simple" 1:1 migration to the new API or if more significant changes are needed because Twitch might've changed or even cut some functionality.

So uh, @anxdpanic if you are out there: Help, please?

**Comment by anxdpanic** [comment]
It's absolutely on my radar, we will likely lose some functionality. But I will be trying to get some of the migration underway.

**Comment by Lookens** [comment]
Same error for me, i search a solution in this post https://github.com/anxdpanic/plugin.video.twitch/issues/537#issuecomment-758930864 but nothing works.

**Comment by apo86** [comment]
That's good to hear, thanks! If there's anything I can do to help please let me know. I'm only lacking competence, not motivation ;)

**Comment by anxdpanic** [comment]
@apo86 or anyone else looking to help, or follow, 
I'll be tracking the updates on https://github.com/anxdpanic/plugin.video.twitch/tree/helix_api

The changes to https://github.com/anxdpanic/script.module.python.twitch won't be breaking, so those can be tracked in master

**Comment by anxdpanic** [comment]
I'd estimate about half way there, it looks like we'll probably lose blacklist (at least until it can be revisited in the future), but may get to add the chat back using IRC. Will find out for sure as I get further.

**Comment by neo-neo1** [comment]
Looks like we’re screwed currently. Twitch v5 API is currently in a planned 4 hour shutdown block with many more planned rolling shutdowns until the final death day Feb 22.

**Comment by anxdpanic** [comment]
That's why I'm trying to keep everyone posted here. Once something usable for testing is ready I will post everything needed as well.

**Comment by neo-neo1** [comment]
> That's why I'm trying to keep everyone posted here. Once something usable for testing is ready I will post everything needed as well.

Thanks for your work. I hope to see Kodi 18 support continued since many still can not run Kodi 19. LibreELEC is a big one as they have yet to add Kodi 19 support to Rasp Pi 3B and many other popular boards. Thanks again.

**Comment by anxdpanic** [comment]
No reason to change Kodi compatibility at this moment, so the helix update will be for the same versions of Kodi.

**Comment by apo86** [comment]
> I'd estimate about half way there, it looks like we'll probably lose blacklist (at least until it can be revisited in the future), but may get to add the chat back using IRC. Will find out for sure as I get further.

Sounds promising, thanks for your efforts!

**Comment by kcs32787github** [comment]
> I'd estimate about half way there, it looks like we'll probably lose blacklist (at least until it can be revisited in the future), but may get to add the chat back using IRC. Will find out for sure as I get further.

Thanks for all you do with this addon. I would be completely lost without it :)

**Comment by MrSprigster** [comment]
@anxdpanic Thank you for your continued efforts, it is much appreciated by a silent many, myself included.  

Sprig

**Comment by xenius** [comment]
Yeah having API issues. Does this mean i cannot use the addon anymore until the addon has an update? Or will the api come back online?

**Comment by MrSprigster** [comment]
For anyone currently experiencing add-on issues:

Twitch API v5 phased shutdown schedule:

February 7, 2022 – 1-hour shutdown starting at 11:00am PDT
February 9, 2022 – 2-hour shutdown starting at 11:00am PDT
February 14, 2022 – 4-hour shutdown starting at 11:00am PDT
February 16, 2022 – 6-hour shutdown starting at 11:00am PDT
February 22, 2022 – 24-hour shutdown starting at 11:00am PDT
February 28, 2022 – v5 permanent shutdown

**Comment by pitsi** [comment]
@MrSprigster 
Thank you very much for posting the schedule. 11am pdt is 9pm for me here at gmt+2, which means that it stops working just when my day is over and I lay down and watch some twitch to relax. And I am probably sleeping when it comes back up.
As it seems, next Tuesday, February 22nd, will be a tough day for me :P 

@anxdpanic 
Can you please upload any development version of the addon in your "twitch development repository"? Alpha, beta, anything is welcome as long as it can play vods or live streams!

**Comment by anxdpanic** [comment]
@pitsi 
That's the plan, and I do _hope_ to have something up before the outage tomorrow. It might only be the 'Following' section but should be usable.

**Comment by pitsi** [comment]
Oh, there is an outage planned for tomorrow. I forgot about it.
I do not have a twitch account, thus I have no use for the "following" section. I just check games or streamers, which I keep as favorites inside kodi.

**Comment by anxdpanic** [comment]
Favorited streams should work in that case, I have followed stuff 'working' and didn't have to change the playback route, so anything favorite pointing to something playable should work once I have something up.

**Comment by pitsi** [comment]
Thank you for the clarification.
What if the favorite does not link to something directly playable but to a list instead, e.g. the streams of a game or the vods of a streamer?

**Comment by anxdpanic** [comment]
That will be hit or miss, depends on how much I can get done today before I push the alpha 👍 
edit: Some of those favorites will have to be recreated though, some routes have already changed due to the api changes.

**Comment by pitsi** [comment]
Thank you :)
God I hate api changes (regardless of platform)!

**Comment by anxdpanic** [comment]
Not sure if I'll get back to it today, but I have it 90%+ working but completely broken on Kodi 19+. 
Hopefully later tonight or tomorrow morning I'll be able to get that working and push a version to the test repositories.
Thanks for your patience 😄

**Comment by anxdpanic** [comment]
Well it looks like it was my portable v19 install that was the issue. 

Zips can be found here: (install python.twitch first)
https://github.com/anxdpanic/script.module.python.twitch/releases/tag/2.0.21-dev
https://github.com/anxdpanic/plugin.video.twitch/releases/tag/2.6.0-dev

I'll work on getting them pushed to the development repository next.

**Comment by anxdpanic** [comment]
Development repository should contain the latest alpha versions. 

If there are any issues with the alphas please open a new issue so it's easier to track.

We can continue with everything else in this thread

😄

**Comment by Lookens** [comment]
Thanks for you work, everything is good for me now !

**Comment by kevinackroyd** [comment]
Working fine for me too, thanks for all your amazing work.
BTW, Anyone wanting to login will need to generate a new OAuth token using the link on the settings page.

**Comment by apo86** [comment]
Thanks, amazing work! All the main functionality looks good to me.

**Comment by MrSprigster** [comment]
@anxdpanic Legend. Thanks again!

**Comment by jonisb** [comment]
Thanks for the updates
I tried a live stream and a VOD both worked, there are some details like VOD length that are incorrect but I'm guessing they are not finished yet.

Very much appreciated. 👍

**Comment by neo-neo1** [comment]
Working great here. While you’re under the hood please consider adding [Add to Queue](https://github.com/anxdpanic/plugin.video.twitch/issues/603) feature. If possible would be greatly appreciated. Thanks

**Comment by kcs32787github** [comment]
3/3 on my installs. Working perfectly! Thanks for the quick work and getting up to date before the death of v5 api!

**Comment by bjoerntw** [comment]
Works great, thank you very much for your effort!

**Comment by rjousse18** [comment]
not working for me. Installed alpha on rpi4 libreelec 10.0.1 (kodi 19.3) : 

2022-02-16 23:29:18.946 T:1569  WARNING <general>: script.module.python.twitch: API version |V5| is deprecated, update to |Helix| by |TBD|
2022-02-16 23:29:19.373 T:1569    ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'AttributeError'>
                                                   Error Contents: module 'twitch.oauth.helix.scopes' has no attribute 'user_read_follows'
                                                   Traceback (most recent call last):
                                                     File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/addon_runner.py", line 13, in <module>
                                                       from twitch_addon import router
                                                     File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/router.py", line 14, in <module>
                                                       from .addon import api
                                                     File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/api.py", line 14, in <module>
                                                       from . import cache, utils
                                                     File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/utils.py", line 24, in <module>
                                                       from .constants import CLIENT_ID, REDIRECT_URI, LIVE_PREVIEW_TEMPLATE, Images, ADDON_DATA_DIR, COLORS, Keys
                                                     File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/constants.py", line 208, in <module>
                                                       SCOPES = [scopes.user_read_follows, scopes.user_edit_follows, scopes.user_read_subscriptions,
                                                   AttributeError: module 'twitch.oauth.helix.scopes' has no attribute 'user_read_follows'
                                                   -->End of Python script error report<--

2022-02-16 23:29:20.215 T:1569     INFO <general>: Python interpreter stopped
2022-02-16 23:29:20.223 T:1568    ERROR <general>: GetDirectory - Error getting plugin://plugin.video.twitch/
2022-02-16 23:29:20.233 T:1015    ERROR <general>: CGUIMediaWindow::GetDirectory(plugin://plugin.video.twitch/) failed

**Comment by ghost** [comment]
> not working for me. Installed alpha on rpi4 libreelec 10.0.1 (kodi 19.3) :
> 
> 2022-02-16 23:29:18.946 T:1569 WARNING : script.module.python.twitch: API version |V5| is deprecated, update to |Helix| by |TBD| 2022-02-16 23:29:19.373 T:1569 ERROR : EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<-- - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS! Error Type: <class 'AttributeError'> Error Contents: module 'twitch.oauth.helix.scopes' has no attribute 'user_read_follows' Traceback (most recent call last): File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/addon_runner.py", line 13, in from twitch_addon import router File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/router.py", line 14, in from .addon import api File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/api.py", line 14, in from . import cache, utils File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/utils.py", line 24, in from .constants import CLIENT_ID, REDIRECT_URI, LIVE_PREVIEW_TEMPLATE, Images, ADDON_DATA_DIR, COLORS, Keys File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/constants.py", line 208, in SCOPES = [scopes.user_read_follows, scopes.user_edit_follows, scopes.user_read_subscriptions, AttributeError: module 'twitch.oauth.helix.scopes' has no attribute 'user_read_follows' -->End of Python script error report<--
> 
> 2022-02-16 23:29:20.215 T:1569 INFO : Python interpreter stopped 2022-02-16 23:29:20.223 T:1568 ERROR : GetDirectory - Error getting plugin://plugin.video.twitch/ 2022-02-16 23:29:20.233 T:1015 ERROR : CGUIMediaWindow::GetDirectory(plugin://plugin.video.twitch/) failed

FYI I get this error when I do not have the dev version of script.module.python.twitch addon installed (the dev version is linked above but also available in the repo through the video plugin dependencies). the video plugin does not seem to depend on the newer script version.

**Comment by anxdpanic** [comment]
I've pushed alpha2 to the release and dev repo. 

- fixes toggle follow
- fixes video durations @jonisb thx for the heads up
- fixes live notification thumbnails
- main menu has been adjusted to accommodate the loss of so many sub-menus
- added 'Add to queue' to playable content @neo-neo1 
- main menu hidden by oauth token presence. an oauth token is required for all the requests

Regarding the oauth token being required, the helix api seems to be designed with only hosted platforms or authenticated users in mind. Since we're not a hosted platform, I don't currently see a way to resolve that otherwise. Wish I had better news there, it's my least favorite outcome of this api change.

**Comment by anxdpanic** [comment]
#622 is the only currently outstanding issue that I'm aware of.
Please report anything you think might be relevant, because it might have just been something I've overlooked or forgotten.

**Comment by apo86** [comment]
> #622 is the only currently outstanding issue that I'm aware of. Please report anything you think might be relevant, because it might have just been something I've overlooked or forgotten.

Are Clips supposed to be working? When I was trying different things earlier I always got errors for Clips (no matter if it was on followed channels, live channels, offline channels)
If this is not a known issue I can write up a proper report tomorrow after work.

**Comment by anxdpanic** [comment]
@apo86 
They are working in my quick tests, there was a spot or two where they needed to be removed from (should be now)

**Comment by Guerretto** [comment]
Thanks for this, man!
Not sure how to get access to the dev repo to install the alpha2. Can someone help? (Raspi4 using LibreElec 10.0.1)

**Comment by anxdpanic** [comment]
@Guerretto 
https://panicked.xyz/repositories/repository.twitch_development-1.0.0.zip (if you're using Kodi 18 or less)
https://panicked.xyz/repositories/matrix/repository.twitch_development-1.0.0+matrix.1.zip (if you're using Kodi 19+)

**Comment by Guerretto** [comment]
Definitely works. Thanks again, this is amazing

**Comment by jonisb** [comment]
Updated to alpha2 on Kodi 18.8
All VODs seem to have "Unknown game" in the description.
I show the currently live channels in a widget and now all channels have a gray thumbnail.

**Comment by mrandersonsmith** [comment]
> Thanks for this, man! Not sure how to get access to the dev repo to install the alpha2. Can someone help? (Raspi4 using > LibreElec 10.0.1)  

  

I sshed into my RaspPI box and used wget \<direct link to zip file\>
to download, then within Kodi addons --> install from zip file. If you dont know what SSH is then you can just copy the zip files to a USB key and then install from Zip file and browse to folder to install. dont use the +matrix version.

**Comment by anxdpanic** [comment]
@jonisb 
Open a new issue with a log for the widget issue when you have time. 
The Unknown Game is something I completely forgot to go back to, I will see if that can be resolved.

**Comment by pitsi** [comment]
I hate to be the party pooper here, but nothing works for me after the upgrade to 2.6.0-alpha1/2.
Bookmarked vods do not work, bookmarked channels do not work, bookmarked games do not work, search does not work and they all show a similar error in the log.
I completely deleted  `~/.kodi/userdata/addon_data/plugin.video.twitch` after the update to the alphas (I saved a backup just in case) so that the addon will start with fresh settings, but the results are the same. Kodi 18.9 on libreelec 9.2.6 x64.

I will definitely open an issue with the relevant log part later in the day. However, I may have to return to the previous version until it is completely fixed.

p.s. As a test, I opened a live stream 5 minutes before the scheduled hour of shutdown and it ran for 1+ hour with no issues at all. I probably got the update to alpha1 after that.

**Comment by neo-neo1** [comment]
Working beautifully! Loving the new Add to Queue. Now I can channel surf live channels easy. Super!

**Comment by anxdpanic** [comment]
@pitsi 
I was incorrect in my original statement, that they would work. All endpoints require an oauth or app token, and app token's won't be a working solution for us. It's rate limited, and there is no way to limit the rate in our situation.

> I've pushed alpha2 to the release and dev repo.
...
> Regarding the oauth token being required, the helix api seems to be designed with only hosted platforms or authenticated users in mind. Since we're not a hosted platform, I don't currently see a way to resolve that otherwise. Wish I had better news there, it's my least favorite outcome of this api change.

**Comment by pitsi** [comment]
Ok, no problem with that. I can make new bookmarks, but for me to do so, search (at the very least) has to work!
On my end, searching for anything makes kodi wait and wait until I press backspace. Shall I open a new issue about saerch only? 
Also, settings > main menu seems to have changed. I have these on my main menu (sceenshot from main pc running kodi 19.3 on linux and twitch 2.5.12), but I can't make it look the same on the system with 2.6-alpha2.

![screenshot00000](https://user-images.githubusercontent.com/5880037/154420915-a88d1967-4abc-4b8b-b0d2-751cd73251aa.png)

Last but not least, I just got a the "an oauth token is required" message. I followed the instructions shown on the message, but nothing happens. I deleted the folder in addon_data once more, but the message does not go away.
For that reason, I can't post any any log output, because I can no longer use search or games menu to make it produce an error.

I will downgrade to 2.5.12 and restore my backup in like ~1 hour if I can't get any logs, because twitch for me is the core reason I use kodi :D

---edit
I now noticed the quoted part of your reply. Does it mean that now everyone has to use credentials and make a new oauth key?

**Comment by neo-neo1** [comment]
I noticed `Attempt reconnection to live streams` option not working when enabled. If a live stream drops it does not reconnect (nor attempt to reconnect).

**Comment by brian-smith-github** [comment]
Thanks, alpha2 working for me too.

**Comment by Guerretto** [comment]
> > Thanks for this, man! Not sure how to get access to the dev repo to install the alpha2. Can someone help? (Raspi4 using > LibreElec 10.0.1)
> 
> I sshed into my RaspPI box and used wget <direct link to zip file> to download, then within Kodi addons --> install from zip file. If you dont know what SSH is then you can just copy the zip files to a USB key and then install from Zip file and browse to folder to install. dont use the +matrix version.

Thanks! I downloaded the file with my computer and copied it to the raspberry through smb, same result.

**Comment by pitsi** [comment]
What is the version of script.module.python.twitch that matches twitch 2.6 alpha 2?

**Comment by brian-smith-github** [comment]
I'm using it with  script.module.python.twitch-2.0.21.alpha1+matrix.1.zip and it's fine, need to use oauth key with it now though.

**Comment by rjousse18** [comment]
I saw in the rights of the api permissions concerning the chat, will we be able to see the chat and send messages soon?

**Comment by pitsi** [comment]
> I'm using it with script.module.python.twitch-2.0.21.alpha1+matrix.1.zip and it's fine, need to use oauth key with it now though.

I though so! Thank you for making it clear. How did you install it though? Here, kodi seems to keep the existing versoin of that dependency and there is no way to install it. Libreelec 10 x64 with kodi 19.3.

**Comment by rjousse18** [comment]
> > I'm using it with script.module.python.twitch-2.0.21.alpha1+matrix.1.zip and it's fine, need to use oauth key with it now though.
> 
> I though so! Thank you for making it clear. How did you install it though? Here, kodi seems to keep the existing versoin of that dependency and there is no way to install it. Libreelec 10 x64 with kodi 19.3.

https://github.com/anxdpanic/script.module.python.twitch/releases

**Comment by pitsi** [comment]
Installing from zip is a solution. However, I think it would be better if the newer version that is required was mentioned in twitch's addon.xml, like so
` <import addon="script.module.python.twitch" version="2.0.20"/>`
As I mentioned above, I use the development repo for updating and this (= not having to install a dev version by hand) is one of the reasons I do so. No offence :)

**Comment by mklcf** [comment]
Quick question: does turbo work in the alpha? And how/where do I find that token in a browser? i did it once, but I completely forgot.

**Comment by Guerretto** [comment]
@mklcf In the addon settings you’ll find a “get token” button that will give you a twitchaddon.page.link/XXXX link (XXXX will be something else) Go there, login with your twitch user and you’ll get the token.

**Comment by anxdpanic** [comment]
@mklcf https://github.com/anxdpanic/plugin.video.twitch/wiki/Private-API-Credentials---OAuth-Token

**Comment by adanrg** [comment]
It seems, that subscriber only VODs are not playable. This is the related snippet from debug log (sorry for the long paste, it's very verbose, but everything is included for error tracing):
... [log snippet removed by anxdpanic]
Any other non subscriber only VOD is playing just fine for me -> great work!

**Comment by anxdpanic** [comment]
@adanrg 
https://github.com/anxdpanic/plugin.video.twitch/wiki/Subscription-and-Turbo-benefits

**Comment by anxdpanic** [comment]
I've posted beta1 to release and repo,

- fixup python.twitch depend version @pitsi (i know for sure i updated it before, not sure where it got undone)
- removed unknown game from videos @jonisb (the api doesn't return this information anymore)
- fixup fanart and icon paths
- fixup error handling, handle unauthorized exceptions to notify users of oauth token requirement
- fixup some dialog formatting
- fixup playing stream thumbnail/preview
- fixup attempt to reconnect on disconnection @neo-neo1 

Thanks again everyone for your feedback and helping get the add-on back in shape 👍

**Comment by anxdpanic** [comment]
@pitsi 
Yes, all users require oauth tokens/accounts to access anything in the add-on now. ☹️

**Comment by jmraker** [comment]
FYI: Where you enter the OAuth token is on the FIRST item in the login tab and not the 2nd one which is really for a private oauth token.  I made that mistake yesterday and alpha1 didn't say anything about an invalid token.  I think I must have used it before with a private key.

The only problem I could find so far in alpha2 is when searching, the game box art icon is low-resolution ("Mario Maker 2" for example).  Maybe it's always been like that.

**Comment by anxdpanic** [comment]
@MrSprigster Thanks for taking care of notifying the forum!

@jmraker I will take a look at the game art work.

**Comment by adanrg** [comment]
> @adanrg https://github.com/anxdpanic/plugin.video.twitch/wiki/Subscription-and-Turbo-benefits

Yep, that was it. Thanks for the help / quick reaction!
After going with the alpha version, I just had to grab a new private OAuth token.

**Comment by pitsi** [comment]
> @pitsi Yes, all users require oauth tokens/accounts to access anything in the add-on now. frowning_face

Great. Another useless account I have to create. Not to mention the entire procedure described in the wiki so as to get the oauth token...

**Comment by anxdpanic** [comment]
@pitsi 
That wiki oauth token is only required if you're using subscriptions. which i'm guessing you aren't if it's a throwaway acct.
You only need the standard oauth key.

**Comment by apo86** [comment]
Is IRC chat already supposed to work? I can enable the setting and I don't get any errors, but I also don't see any chat anywhere.
I don't remember ever using that feature. Was there another addon dependency for it?

**Comment by anxdpanic** [comment]
@apo86 The  IRC add-on is only available in older Kodi's. I haven't tried it yet, and will probably have to have that script updated.

**Comment by anxdpanic** [comment]
I've pushed beta2 to release and repo
- fixup game boxart resolutions @jmraker 
- fixup clip browsing when showing game in title
- fixup video and stream fanart

Sorry for the back to back releases, stepping away and not sure I'll be back to it today so I wanted to get the fixes that were complete out asap.

**Comment by anxdpanic** [comment]
I've pushed beta3 to release and repo along with alpha2 of python.twitch (beta3 will still work with alpha1 of python.twitch)

- fixup rerun playback error
- more error handling fixups

**Comment by neo-neo1** [comment]
Using beta2 I’ve noticed periodically while watching a stream I would get “Currently Live” notifications but it lists every single stream I’m following as if I just went online or restarted Kodi. This happens while watching an active stream. `Group notifications at start` and `notify at start` options enabled.

**Comment by anxdpanic** [comment]
Will look at that next @neo-neo1, likely tomorrow morning sometime. 👍

**Comment by pepszzi** [comment]
Thanks for your hard work @anxdpanic !
Since beta3 no VODs are running. I get an error that oauth key is missing, but it’s there. Live-Streams ok. Got back to alpha1 now and VODs working again.

**Comment by neo-neo1** [comment]
> Thanks for your hard work @anxdpanic !
> Since beta3 no VODs are running. I get an error that oauth key is missing, but it’s there. Live-Streams ok. Got back to alpha1 now and VODs working again.


Can confirm the above issue. Same here.

**Comment by anxdpanic** [comment]
@pepszzi and @neo-neo1 
I pushed the fixed as beta4, that was completely my fault.

**Comment by anxdpanic** [comment]
I've pushed beta5
- fixup search streams with title formats which included viewers
- cleanup title formatting when game or viewers is missing

**Comment by pitsi** [comment]
Now that I know an oauth key will be mandatory from March 1st, I switched back to 2.5.x and I am waiting for the inevitable to happen. 

I disabled the development repo, but I kept it installed for future usage. The problem is that on every single launch I get a notification that twitch and python-twitch (script.module.python.twitch) were updated AND then they show up as available for update in settings > addons! Twitch wants to be updated to 2.6.0 alpha2 and python-twitch to 2.0.21 alpha1, which are not even the latest versions on the repo! 
I have left auto-update to on for both if it helps. Kodi 18.9 on libreelec 9.2.6 x64.

---edit
I forgot to say that I removed and reinstalled both, taking them from the main repo, but the weird thing above persists. It also removed SOME entries from my favorites, e.g. only one of the two games and two of the three streamers I had bookmarked were removed, not everything.

**Comment by anxdpanic** [comment]
@neo-neo1 
I haven't been able to reproduce your issue with live notifications, if it happens again can you create a new issue with the debug log. 👍 

@pitsi 
I can see that being frustrating. Kodi is probably seeing the version of the add-on from the `addons/packages/` folder where Kodi downloads the zips to.

---

I've pushed beta6, 
- add private credential validation
  - if the private oauth token shows the add-ons client id (it shouldn't) and the add-on oauth token is empty, the add-on will move the private oauth token to the correct setting instead of causing popups.
  - if playback or any function that uses private oauth credentials fails, a dialog will notify of the issue instead of failing with a generic exception

I'm not aware of any outstanding issues that I'm able to resolve. If there is anything, please let me know here or with a new issue.
I'm targeting submission to the official repository for Monday, to allow some time for submission review before the end of the month.

**Comment by neo-neo1** [comment]
> @neo-neo1 I haven't been able to reproduce your issue with live notifications, if it happens again can you create a new issue with the debug log. 👍
>

So far no issues in the latest beta. All good now 👍

**Comment by neo-neo1** [comment]
Just updated to beta6 and `Following Live Channels` Menu errors out and doesn’t work?

**Comment by neo-neo1** [comment]
Had an initial issue with beta6 but solved it by uninstalling and installing. Hope this helps someone.

**Comment by pitsi** [comment]
> I can see that being frustrating. Kodi is probably seeing the version of the add-on from the `addons/packages/` folder where Kodi downloads the zips to.

I thought about it too, but the `~/.kodi/addons/pacages` folder is cleaned on every boot via autostart.sh
```
# cat .config/autostart.sh | grep packages
rm /storage/.kodi/addons/packages/*.zip
```
And oddly enough, the versions that are re-downloaded in there on each boot are the ones that come from the main repo.
```
# ls .kodi/addons/packages/
plugin.video.twitch-2.5.12.zip          script.module.python.twitch-2.0.19.zip
```
I will post at the thread on kodi's forum, in case some dev can find the reason.

**Comment by anxdpanic** [comment]
I've submitted the add-ons to the official repository for review. Thank you everyone for your testing, feedback and patience!

**Comment by fragande** [comment]
> I've submitted the add-ons to the official repository for review. Thank you everyone for your testing, feedback and patience!

Thank _you_ for your continued efforts 🙇‍♂️

**Comment by apo86** [comment]
> I've submitted the add-ons to the official repository for review. Thank you everyone for your testing, feedback and patience!

Thank you for everything, we would've been completely lost without you!

**Comment by cmsimike** [comment]
Thank you!

**Comment by cmsimike** [comment]
This might be a problem on my end, but I am trying to manually update the Twitch plugin. The version in matrix says that it is `2.6.0+matrix.1` but when I load the addon I get an exception
```
2022-02-22 15:27:58.552 T:7181     INFO <general>: initializing python engine.
2022-02-22 15:27:58.603 T:7153     INFO <general>: Python interpreter stopped
2022-02-22 15:27:58.923 T:7181  WARNING <general>: script.module.python.twitch: API version |V5| is deprecated, update to |Helix| by |TBD|
2022-02-22 15:27:59.260 T:7181    ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'AttributeError'>
                                                   Error Contents: module 'twitch.oauth.helix.scopes' has no attribute 'user_read_follows'
                                                   Traceback (most recent call last):
                                                     File "/home/mike/.kodi/addons/plugin.video.twitch/resources/lib/addon_runner.py", line 13, in <module>
                                                       from twitch_addon import router
                                                     File "/home/mike/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/router.py", line 14, in <module>
                                                       from .addon import api
                                                     File "/home/mike/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/api.py", line 15, in <module>
                                                       from . import cache, utils
                                                     File "/home/mike/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/utils.py", line 24, in <module>
                                                       from .constants import CLIENT_ID, REDIRECT_URI, LIVE_PREVIEW_TEMPLATE, Images, ADDON_DATA_DIR, COLORS, Keys
                                                     File "/home/mike/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/constants.py", line 208, in <module>
                                                       SCOPES = [scopes.user_read_follows, scopes.user_edit_follows, scopes.user_read_subscriptions,
                                                   AttributeError: module 'twitch.oauth.helix.scopes' has no attribute 'user_read_follows'
                                                   -->End of Python script error report<--

```

I tried clearing out my addon data, but I am still getting this error. Thoughts?

**Comment by anxdpanic** [comment]
@cmsimike https://github.com/anxdpanic/script.module.python.twitch/releases/tag/v2.0.21

**Comment by cmsimike** [comment]
@anxdpanic Thank you - just figured it out and came here to update the thread. Indeed this was the issue. Forgot there was this script that needed to be updated. THANKS AGAIN!

**Comment by Guerretto** [comment]
Neither on Raspi4

**Comment by beigexperience** [comment]
Thanks for the timely update. I just updated the addon from the repo and it works fine on a raspi3b running libreelec

**Comment by Guerretto** [comment]
Confirmed, works again

**Comment by pitsi** [comment]
Happening now on my end. Main pc running kodi 19.3 on debian testing/unstable  x64 and twitch 2.5.12 from the main repo (= the dev repo is not installed here). 
```
2022-02-25 11:18:43.283 T:3502    ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'ModuleNotFoundError'>
                                                   Error Contents: No module named 'twitch.oauth.v5'
                                                   Traceback (most recent call last):
                                                     File "/home/user/.kodi/addons/plugin.video.twitch/resources/lib/service_runner.py", line 12, in <module>
                                                       from twitch_addon import service
                                                     File "/home/user/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/service.py", line 21, in <module>
                                                       from .addon.constants import Keys
                                                     File "/home/user/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/constants.py", line 14, in <module>
                                                       from twitch.oauth.v5 import scopes
                                                   ModuleNotFoundError: No module named 'twitch.oauth.v5'
                                                   -->End of Python script error report<--
```
If it helps. I got a minor update on python3 3.9.10 (3.9.10-1 to 3.9.10-2) and this tiny bit is mentioned in the changelog.
`  * Make test suite support Expat >=2.4.5.`

If it is an api issue, please tell me. If it is a python issue, ignore the above. I am now used to addons breaking because of the python updates of my distro.
No issues on the libreelec 9.2.6 x64 installation that has the same version of twitch... and python2.

**Comment by anxdpanic** [comment]
@pitsi script.module.python.twitch has been reviewed and merged to official repo first (as required), so you'll have to rollback the python.twitch version.

**Comment by pitsi** [comment]
You are right, python-twitch was upgraded to 2.0.21 under my nose :P 
```
$ cat .kodi/addons/*.twitch/addon.xml | grep "addon id"
<addon id="plugin.video.twitch" version="2.5.12+matrix.1" name="Twitch" provider-name="anxdpanic, A Talented Community">
<addon id="script.module.python.twitch" name="python-twitch for Kodi" version="2.0.21+matrix.1" provider-name="anxdpanic, A Talented Community">
```
I will do the downgrade later in the day. If you have to move twitch 2.6.0 to the main repo, please do so.

**Comment by Slizbury** [comment]
Upgrading to the latest version of the Twitch plugin (2.6.0+matrix.1) has restored functionality of the plugin for me, but the "Browse" option is now missing.  I only see my "Following" people and a "Search" option, but the "Browse" option (and I think maybe one or two others) is gone.  Hopefully that can come back at some point.

I know this is all in mid-development with the API switch, and I appreciate the effort that has been taken to get things working with the new API!  Just wanted to give some feedback on what I was seeing after my upgrade.

**Comment by THeXES** [comment]
When I put the link I get "Dynamic Link Not Found", is there any solution to this?

**Comment by MrSprigster** [comment]
The link is case sensitive.  Included below for your convenience.

https://twitchaddon.page.link/Z1iP

**Comment by M-Reimer** [comment]
If I try to click "Get OAuth token", then a message pops up telling "Failed to connect to Twitch API" and I have this in my logs:

```
2022-02-26 13:25:48.972 T:537      INFO <general>: initializing python engine.
2022-02-26 13:25:50.032 T:537     ERROR <general>: Traceback (most recent call last):
                                                   
2022-02-26 13:25:50.032 T:537     ERROR <general>: 
2022-02-26 13:25:50.032 T:537     ERROR <general>:   File "/var/lib/kodi/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/router.py", line 290, in run
                                                       twitch_api = api.Twitch()
                                                   
2022-02-26 13:25:50.032 T:537     ERROR <general>: 
2022-02-26 13:25:50.032 T:537     ERROR <general>:   File "/var/lib/kodi/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/api.py", line 54, in __init__
                                                       if not self.valid_token(self.client_id, self.access_token, self.required_scopes):
                                                   
2022-02-26 13:25:50.032 T:537     ERROR <general>: 
2022-02-26 13:25:50.032 T:537     ERROR <general>:   File "/var/lib/kodi/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/common/cache.py", line 104, in memoizer
                                                       result = func(*args, **kwargs)
                                                   
2022-02-26 13:25:50.032 T:537     ERROR <general>: 
2022-02-26 13:25:50.032 T:537     ERROR <general>:   File "/var/lib/kodi/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/api.py", line 62, in valid_token
                                                       if token_check['client_id'] == self.client_id:
                                                   
2022-02-26 13:25:50.032 T:537     ERROR <general>: 
2022-02-26 13:25:50.032 T:537     ERROR <general>: KeyError: 'client_id'
                                                   
2022-02-26 13:25:50.032 T:537     ERROR <general>: 
2022-02-26 13:25:50.032 T:537     ERROR <general>: Twitch: None
2022-02-26 13:25:50.032 T:537      INFO <general>: CPythonInvoker(9, /var/lib/kodi/.kodi/addons/plugin.video.twitch/resources/lib/addon_runner.py): script successfully run
2022-02-26 13:25:50.059 T:537      INFO <general>: Python interpreter stopped
```
Edit: This happens if you once entered an invalid OAuth token. The only way to enter a new one is to use "Defaults" in the settings to get rid of the invalid token first.

Works for me now but I had to manually install. Does not come from the addon repo, so far. Thank you for the fix!

**Comment by pitsi** [comment]
To the ones that do NOT use the dev repo and get the updates from the main one only.
How does your twitch 2.5.12 cope with the new python-twitch 2.0.21 (because these are the latest versions provided from the main repo)?
Mine pops up the error above and downgrading python-twitch to make it work (and keep it downgraded until March 1st) is a pita.

**Comment by QMaxy** [comment]
> How does your twitch 2.5.12 cope with the new python-twitch 2.0.21 (because these are the latest versions provided from the main repo)?

Using only the main repo I've received python-twich 2.0.21 update, but Twich add-on till remain to 2.5.12 release. This combination doesn't work with pop-up error everytime Kodi starts.
I tried manually downgrading python-twich to 2.0.19 and all works fine, but the next time I started Kodi, python-twich has been auto-upgraded to .21 and, again it doesn't work.
Next, I downloaded Twich add-on 2.6.0 from here (not yet available from main depo), obtained my OAuthed token, and now all seems working fine.

**Comment by pitsi** [comment]
Same thing here then. 

@anxdpanic 
Now that api v5 is dead, along with twitch 2.5,  can you please push twitch 2.6 to the main repo?

**Comment by anxdpanic** [comment]
@pitsi It was submitted to the official repository ~8days ago. It needed to be reviewed by another Team member when they have time before it could be merged.

Looks like it's been reviewed and merged on matrix, lower versions will follow later today.

**Comment by pitsi** [comment]
Yea, I remember you saying about submitting it. I checked like ~1 hour ago and both leia and matrix were on 2.5.x. Now matrix is on 2.6.0.

---edit
I finally got it to work again! The good news is that I did not have to reset settings or remove anything at all, so all my favorites are still there and working! The bad news is that when selecting "get oauth token" or "oauth token" below it in my kodi 19.3, the entire kodi crashes. I blame my distro's python 3.9.x for that because it happens with other addons too. I will post a log when I have time. 
I managed to put it there by using "send text to kodi" through the web interface. Kodi crashed again, but at least the oauth key was stored.

**Comment by anxdpanic** [comment]
Everything has been merged to the official repository now. I'll close this issue now in favor of specific issues encountered with the new version.

Thanks again everyone!

**Comment by fossdd** [comment]
Thank you guys, I can verify the latest version works perfectly!

**Comment by dgiglio** [comment]
Hi @anxdpanic 
first of all, thanks for your work. I'm encountering this issue:
I can watch Twitch on Kodi 19.x using the 2.6 plugin version on a AMD64 architecture.
I cannot do the same on my Raspberry4 with a 32bit Debian, since the official Kodi repo still provides the deprecated 2.5.x plugin version only.
How can I force an upgrade? Can I proceed installing directly the zip?
Thanks in advance
